src/v/cluster/config_frontend.h

```diff
@@ -0,0 +1,53 @@
+/*
+ * Copyright 2021 Vectorized, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+
+#pragma once
+
+#include "cluster/controller_stm.h"
+
+namespace cluster {
+
+struct config_update final {
+    std::vector<std::pair<ss::sstring, ss::sstring>> upsert;
+    std::vector<ss::sstring> remove;
+};
+
+class config_frontend final {
+public:
+    static constexpr ss::shard_id version_shard = 0;
+
+    config_frontend(
+      ss::sharded<controller_stm>&, ss::sharded<ss::abort_source>&);
+
+    ss::future<std::error_code>
+    patch(config_update&, model::timeout_clock::time_point);
+
+    ss::future<std::error_code>
+    set_status(config_status&, model::timeout_clock::time_point);
+
+    ss::future<> start();
+    ss::future<> stop();
+
+    void set_next_version(config_version v);
+
+private:
+    ss::sharded<controller_stm>& _stm;
+    ss::sharded<ss::abort_source>& _as;
+
+    // Initially unset, frontend is not writeable until backend finishes
+    // init and calls set_next_version.
+    config_version _next_version{
+      config_version_unset}; // Only maintained on `version_shard`
+
+    // Serialize writes to generate contiguous version numbers.
+    mutex _write_lock;
```

Comment:
My comment shouldn't say contiguous -- really this is just going for a monotonic number that can be known before writing into the log, so that we can use it as the key.  Having a version number in the key should let us come back later and write nulls to those if we want to avoid an unbounded-length store of config deltas (even though I'm not doing that yet).